packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -1445,38 +1445,40 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
           controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
           updatedTypesQuery: Option.none,
         };
-      } else if (isTypeof(left, 'root_fields_changed')) {
-        // compatible migration: some core fields have been updated
-        return {
-          ...stateP,
-          controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
-          // we must "pick-up" all documents on the index (by not providing a query)
-          updatedTypesQuery: Option.none,
-          logs: [
-            ...stateP.logs,
-            {
-              level: 'info',
-              message: `Kibana is performing a compatible upgrade and the mappings of some root fields have been changed. For Elasticsearch to pickup these mappings, all saved objects need to be updated. Updated root fields: ${left.updatedFields}.`,
-            },
-          ],
-        };
-      } else if (isTypeof(left, 'types_changed')) {
-        // compatible migration: some fields have been updated, and they all correspond to SO types
-        const updatedTypesQuery = Option.fromNullable(buildPickupMappingsQuery(left.updatedTypes));
+      } else if (isTypeof(left, 'mappings_changed')) {
+        if (left.updatedTypes?.length) {
+          // compatible migration: the mappings of some SO types have been updated
+          const updatedTypesQuery = Option.fromNullable(
+            buildPickupMappingsQuery(left.updatedTypes)
+          );
 
-        return {
-          ...stateP,
-          controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
-          // we can "pick-up" only the SO types that have changed
-          updatedTypesQuery,
-          logs: [
-            ...stateP.logs,
-            {
-              level: 'info',
-              message: `Kibana is performing a compatible upgrade and NO root fields have been updated. Kibana will update the following SO types so that ES can pickup the updated mappings: ${left.updatedTypes}.`,
-            },
-          ],
-        };
+          return {
+            ...stateP,
+            controlState: 'UPDATE_TARGET_MAPPINGS_PROPERTIES',
+            // we can "pick-up" only the SO types that have changed
+            updatedTypesQuery,
+            logs: [
+              ...stateP.logs,
+              {
+                level: 'info',
+                message: `Documents of the following SO types will be updated, so that ES can pickup the updated mappings: ${left.updatedTypes}.`,
+              },
+            ],
+          };
+        } else {
+          // only core fields have changed (no need to update docs in order to pickup changes in mappings)
```

Comment:
this is primarily because we have `dynamic: strict` in the base mappings for the root fields https://github.com/elastic/kibana/blob/main/packages/core/saved-objects/core-saved-objects-migration-server-internal/src/core/build_active_mappings.ts#L36-L49

So we can't ever have documents with fields in them that haven't previously been mapped.

So if we ever add a new root field mapping it means we're guaranteed that we don't have to "pick up" the fields in existing docs.